src/sequence/mod.rs

```diff
@@ -719,22 +724,368 @@ mod tests {
+            let perms = BTreeMap::default();
+            let owner_op = replica1.set_public_policy(actor1, perms)?;
+            replica2.apply_public_policy_op(owner_op)?;
+
+            let dataset_length = dataset.len() as u64;
+
+            let mut ops = vec![];
+            for data in dataset {
+                    let op = replica1.append(data)?;
+                    ops.push(op);
+            }
+
+            for op in ops.clone() {
+                let delivery_chance : u32 = OsRng.gen_range(0, 10);
+                if delivery_chance < 2 {
+                    replica2.apply_data_op(op)?;
+                }
+            }
+
+            // shere we statistically should have dropped some messages
+            assert_ne!(replica2.len(), replica1.len());
+
+            // reapply all ops
+            for op in ops {
+                    replica2.apply_data_op(op)?;
+            }
+
+            // now we converge
+            verify_data_convergence(vec![replica1, replica2], dataset_length);
+
+        }
+
+
+        #[test]
+        fn proptest_converge_with_shuffled_ops_from_many_while_dropping_some_at_random(
+            dataset in generate_dataset(1000),
+            mut replicas in generate_replicas(100),
+        ) {
+            let dataset_length = dataset.len() as u64;
+
+            // generate an ops set using random replica for each data
+            let mut ops = vec![];
+            for data in dataset {
+                if let Some(replica) = replicas.choose_mut(&mut OsRng)
+                {
+                    let op = replica.append(data)?;
+                    ops.push(op);
+                }
+            }
+
+            let opslen = ops.len() as u64;
+            prop_assert_eq!(dataset_length, opslen);
+
+            // now we randomly shuffle ops and apply at each replica
+            for replica in &mut replicas {
+
+                let mut ops = ops.clone();
+                ops.shuffle(&mut OsRng);
+
+                for op in ops {
+                    let delivery_chance : u32 = OsRng.gen_range(0, 10);
+                    if delivery_chance > 3 {
+                        replica.apply_data_op(op)?;
+                    }
+                }
+            }
+
+            // TODO: Currently this wont converge. Data is missing.
+            // We need to be attempting to apply ops which are known to be not causally ready...
+            // We can get errors there by dropping from sync with one replica...
+            //
+            // Q: how to expand this across many replicas and simulate dropping messages? Is that wortwhile here at the data itself...?
+            verify_data_convergence(replicas, dataset_length);
+        }
+
+        #[test]
+        fn proptest_converge_with_shuffled_ops_including_bad_ops_which_error_and_are_not_applied(
+            dataset in generate_dataset(1000),
+            // should be same number as dataset
+            bogus_dataset in generate_dataset(1000),
```

Comment:
Since you need the same number of entries as in `dataset`, and since the data is not really what makes the ops bogus (but the owner being diff. I think), I think you can just clone the dataset.

Reply:
I was thikning we needed different data in case the same data was somehow recognised as a no-op in the seq itself (through a bug there on the base seq detection)